src/command/reboot.rs

```diff
@@ -0,0 +1,110 @@
+use std::env;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use clap::{Arg, ArgMatches, Command as ClapCommand};
+use futures::future::join_all;
+use tokio::sync::Semaphore;
+
+use crate::error::ColmenaError;
+use crate::job::{JobMonitor, JobState, JobType};
+use crate::nix::NodeFilter;
+use crate::progress::SimpleProgressOutput;
+use crate::util;
+
+pub fn subcommand() -> ClapCommand<'static> {
+    let command = ClapCommand::new("reboot")
+        .about("Reboot remote machines")
+        .trailing_var_arg(true)
+        .arg(
+            Arg::new("parallel")
+                .short('p')
+                .long("parallel")
+                .value_name("LIMIT")
+                .help("Deploy parallelism limit")
+                .long_help(
+                    r#"Limits the maximum number of hosts to reboot in parallel.
+
+In `colmena reboot`, the parallelism limit is disabled (0) by default.
+"#,
+                )
+                .default_value("0")
+                .takes_value(true)
+                .validator(|s| match s.parse::<usize>() {
+                    Ok(_) => Ok(()),
+                    Err(_) => Err(String::from("The value must be a valid number")),
+                }),
+        );
+
+    util::register_selector_args(command)
+}
+
+pub async fn run(_global_args: &ArgMatches, local_args: &ArgMatches) -> Result<(), ColmenaError> {
+    let hive = util::hive_from_args(local_args).await?;
+    let ssh_config = env::var("SSH_CONFIG_FILE").ok().map(PathBuf::from);
+
+    let filter = local_args.value_of("on").map(NodeFilter::new).transpose()?;
+
+    let mut targets = hive.select_nodes(filter, ssh_config, true).await?;
+
+    let parallel_sp = Arc::new({
+        let limit = local_args
+            .value_of("parallel")
+            .unwrap()
+            .parse::<usize>()
+            .unwrap();
+
+        if limit > 0 {
+            Some(Semaphore::new(limit))
+        } else {
+            None
+        }
+    });
+
+    let mut output = SimpleProgressOutput::new(false);
+
+    let (monitor, meta) = JobMonitor::new(output.get_sender());
+    let meta = meta.run(|meta| async move {
+        let mut futures = Vec::new();
+
+        for (name, target) in targets.drain() {
+            let parallel_sp = parallel_sp.clone();
+
+            let mut host = target.into_host().unwrap();
+
+            let job = meta.create_job(JobType::Execute, vec![name.clone()])?;
+
+            futures.push(job.run_waiting(|job| async move {
+                let permit = match parallel_sp.as_ref() {
+                    Some(sp) => Some(sp.acquire().await.unwrap()),
+                    None => None,
+                };
+
+                job.state(JobState::Running)?;
+
+                host.set_job(Some(job));
+                host.run_command(&["reboot"]).await?;
```

Comment:
This will exit abnormally, but I'm not sure how to tell a successful reboot from any other error.

Per man page:
> ssh exits with the exit status of the remote command or with 255 if an error occurred.

Reply:
Is this because of the SSH connection dying due to the reboot? I tested a few times, and `systemctl reboot` seems to fare better, i.e. `ssh root@foobar systemctl reboot` exits with code 0. But the exit code of `ssh root@foobar reboot` is inconsistent for me anyway, so YMMV.

From systemd's man-page:

> reboot
> Shut down and reboot the system. This is mostly equivalent to systemctl start reboot.target --job-mode=replace-irreversibly --no-block,

> --no-block
>     Do not synchronously wait for the requested operation to finish. If this is not specified, the job will be verified, enqueued and systemctl will wait until the unit's
>     start-up is completed. By passing this argument, it is only verified and enqueued. This option may not be combined with --wait.

So my best guess is, "systemctl reboot" exits as fast as possible after submitting the reboot request, and the SSH session has time to end cleanly. But this might not be a guarantee either.

For what it's worth, NixOps also uses `systemctl reboot`. https://github.com/NixOS/nixops/blob/d939c9bcae184d10583bf48211739b5b6521fc64/nixops/backends/__init__.py#L291-L300